docs/testing-with-django.md

```diff
@@ -1,8 +1,237 @@
+                api_result = my_webhook(request)
+
+        self.assertTrue(rate_limit_mock.called)
+
+### Template tests
+
+In [zerver/tests/test_templates.py](https://github.com/zulip/zulip/blob/master/zerver/tests/test_templates.py)
+we have a test that renders all of our back end templates with
+a "dummy" context, to make sure the templates don't have obvious
+errors.  (These tests won't catch all types of errors; they are
+just a first line of defense.)
+
+### SQL performance tests
+
+A common class of bug with Django systems is to handle bulk data in
+an inefficient way, where the back end populates objects for join tables
+with a series of individual queries that give O(N) latency.  (The
+remedy is often just to call `select_related()`, but sometimes it
+requires a more subtle restructuring of the code.)
+
+We try to prevent these bugs in our tests by using a context manager
+called `queries_captured()` that captures the SQL queries used by
+the back end during a particular operation.  We make assertions about
+those queries, often simply asserting that the number of queries is
+below some threshold.
+
+### Event-based tests
+
+The Zulip back end has a mechanism where it will fetch initial data
+for a client from the database, and then it will subsequently apply
+some queued up events to that data to the data structure before notifying
+the client.  The `EventsRegisterTest.do_test()` helper helps tests
+verify that the application of those events via apply_events() produces
+the same data structure as performing an action that generates said event.
+(This is a bit esoteric, but if you read the tests, you will see some of
+the patterns.)
+
+### Negative tests
+
+It is important to verify error handling paths for endpoints, particularly
+situations where we need to ensure that we don't return results to clients
+with improper authentication or with limited authorization.  A typical test
+will call the endpoint with either a non-logged in client, an invalid API
+key, or missing input fields.  Then the test will call `assert_json_error()`
+to verify that the endpoint is properly failing.
+
+## Testing considerations
+
+Here are some things to consider when writing new tests:
+
+- **Duplication** We try to avoid excessive duplication in tests. 
+If you have several tests repeating the same type of test setup,
+consider making a setUp() method or a test helper.
+
+- **Network independence** Our tests should still work if you don't
+have an internet connection.  For third party clients, you can simulate
+their behavior using fixture data.  For third party servers, you can
+typically simulate their behavior using mocks.
+
+- **Coverage** We have 100% line coverage on several of our backend
+modules.  You can use the `--coverage` option to generate coverage
+reports, and new code should have 100% coverage, which generally requires
+testing not only the "happy path" but also error handling code and
+edge cases.
```

Comment:
Basically it's just a webapp for exploring coverage data.  Basically if you visit an individual job page page (e.g. the `test-backend` run) like this, you can nicely browse the coverage data: https://coveralls.io/jobs/17451909